### PR TITLE
Pipeline script

### PR DESCRIPTION
**Pull request by spigo900**
Implements a script to queue all preprocessing steps from one script. The preprocessing steps are queued as separate Slurm jobs with dependencies on each other, set up to cancel immediately if one of their dependencies is canceled/fails. This script can also supports running each step iteratively and interactively by running the Slurm scripts in Bash.

I haven't yet tested this with color segmentation or instance segmentation. Once I have both working I may add segmentation server startup to this.

### CODE REVIEW

**Comment by spigo900** [comment]
I've since tested this with color segmentation and it works as far as I could tell. I did some work here to separate the segmentation and refinement scripts as well which means color and instance segmentation can run in parallel.

As for instance segmentation, I've added code for server startup and shutdown but still need to test it to confirm it works.

**Comment by spigo900** [comment]
Testing is still in progress and should be done today assuming no more bugs to fix. Since the basic setup seems to work, I think this is ready for review now -- though I may push a few bug fixes between today and Monday.

**Comment by spigo900** [comment]
Update: The pipeline finished, and as far as I can tell, everything works.

File: adam/experiment/run_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,646 @@
+"""
+A script intended to run the entire M3 pipeline.
```
Comment (lichtefeld) [inline adam/experiment/run_pipeline.py:2]: Is this supposed to be Phase 3 or milestone 3?

File: adam/experiment/run_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+
+
+def stop_slurm_jobs():
+    """
+    Exit handler to clean up Slurm jobs.
+    """
+    for job_id in slurm_jobs_scheduled:
+        logger.info("Canceling previously scheduled job %s...", job_id)
+        cancel_command = ["scancel", str(job_id)]
+        print(" ".join(shlex.quote(part) for part in cancel_command))
```
Comment (lichtefeld) [inline adam/experiment/run_pipeline.py:45]: Is this a debugging print statement or an intended one? It's unclear to me which its suppose to be.

File: adam/experiment/run_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,646 @@
+"""
+A script intended to run the entire M3 pipeline.
+
+Note that this script does not take care of starting the object segmentation server. You have to do
+that separately.
```
Comment (lichtefeld) [inline adam/experiment/run_pipeline.py:5]: Is this still accurate?

File: slurm/adam.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+#SBATCH --job-name=ADAM
+#SBATCH --account=borrowed
+#SBATCH --partition=scavenge
+#SBATCH --qos=scavenge
```
Comment (lichtefeld) [inline slurm/adam.sh:5]: Do we want to run on `scavenge`? Running on `ephemeral` gives us a guarantee of 1 hour minimum runtime which given we're only requesting an hour for this job seems reasonable?

File: slurm/segment.sh [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+#SBATCH --job-name=adamSegment
+#SBATCH --account=borrowed
+#SBATCH --partition=scavenge
+#SBATCH --qos=scavenge
```
Comment (lichtefeld) [inline slurm/segment.sh:5]: Same note on partition here

**Review by lichtefeld — CHANGES_REQUESTED** [review]
Looks good in general, I haven't tried to make a params file yet but it doesn't seem difficult. 

Just a few minor questions :)

File: adam/experiment/run_pipeline.py (hunk 1 above)
Comment (spigo900) [reply adam/experiment/run_pipeline.py:2]: Phase 3. Whoops.

File: slurm/adam.sh (hunk 4 above)
Comment (spigo900) [reply slurm/adam.sh:5]: I wasn't aware of the guarantee, but that makes sense.

File: adam/experiment/run_pipeline.py (hunk 2 above)
Comment (spigo900) [reply adam/experiment/run_pipeline.py:45]: Actual one, if I recall correctly..

File: adam/experiment/run_pipeline.py (hunk 3 above)
Comment (spigo900) [reply adam/experiment/run_pipeline.py:5]: Nope. 🙃

File: slurm/adam.sh (hunk 4 above)
Comment (lichtefeld) [reply slurm/adam.sh:5]: It's detailed here in the [shared resources policy](https://github.com/isi-vista/saga-cluster/wiki/Adding-a-Project-to-SAGA:-Policy-(Work-in-Progress)#shared-resources)

**Review by lichtefeld — APPROVED** [review]
lgtm!

**Pull request merged**